common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,408 @@
+}
+
+// NewInMemoryStore returns a new InMemoryStore
+func NewInMemoryStore[
+	CHAIN_ID types.ID,
+	ADDR, TX_HASH, BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+](
+	ctx context.Context,
+	lggr logger.SugaredLogger,
+	chainID CHAIN_ID,
+	keyStore txmgrtypes.KeyStore[ADDR, CHAIN_ID, SEQ],
+	txStore txmgrtypes.TxStore[ADDR, CHAIN_ID, TX_HASH, BLOCK_HASH, R, SEQ, FEE],
+	config txmgrtypes.InMemoryStoreConfig,
+) (*InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE], error) {
+	ms := InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		lggr:     lggr,
+		chainID:  chainID,
+		keyStore: keyStore,
+		txStore:  txStore,
+
+		addressStates: map[ADDR]*AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{},
+	}
+
+	maxUnstarted := int(config.MaxQueued())
+	addresses, err := keyStore.EnabledAddressesForChain(ctx, chainID)
+	if err != nil {
+		return nil, fmt.Errorf("new_in_memory_store: %w", err)
+	}
+	for _, fromAddr := range addresses {
+		txs, err := txStore.AllTransactions(ctx, fromAddr, chainID)
+		if err != nil {
+			return nil, fmt.Errorf("address_state: initialization: %w", err)
+		}
+		as, err := NewAddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE](lggr, chainID, fromAddr, maxUnstarted, txs)
+		if err != nil {
+			return nil, fmt.Errorf("new_in_memory_store: %w", err)
+		}
+
+		ms.addressStates[fromAddr] = as
+	}
+
+	return &ms, nil
+}
+
+// CreateTransaction creates a new transaction for a given txRequest.
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CreateTransaction(
+	ctx context.Context,
+	txRequest txmgrtypes.TxRequest[ADDR, TX_HASH],
+	chainID CHAIN_ID,
+) (
+	txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	error,
+) {
+	return txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}, nil
+}
+
+// FindTxWithIdempotencyKey returns a transaction with the given idempotency key
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxWithIdempotencyKey(ctx context.Context, idempotencyKey string, chainID CHAIN_ID) (*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], error) {
+	return nil, nil
+}
+
+// CheckTxQueueCapacity checks if the queue capacity has been reached for a given address
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CheckTxQueueCapacity(ctx context.Context, fromAddress ADDR, maxQueuedTransactions uint64, chainID CHAIN_ID) error {
+	return nil
+}
+
+// FindLatestSequence returns the latest sequence number for a given address
+// It is used to initialize the in-memory sequence map in the broadcaster
+// TODO(jtw): this is until we have a abstracted Sequencer Component which can be used instead
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindLatestSequence(ctx context.Context, fromAddress ADDR, chainID CHAIN_ID) (seq SEQ, err error) {
```

Comment:
Just an fyi (no action needed): This method will probably be removed after the sequence refactor I'm working on. You can still include this in the meantime